### ISSUE TITLE
Behavior change in `incrementMinorIfNotOnRelease` incrementer when on release branch

### ISSUE THREAD

**Issue by mateuszkwiecinski** [allegro/axion-release-plugin#391]
**Context:** 
I'm using `incrementMinorIfNotOnRelease` incrementer by setting 
```
scmVersion.versionIncrementer = PredefinedVersionIncrementer.versionIncrementerFor("incrementMinorIfNotOnRelease")
```

**Repro steps:**
1. Checkout a commit tagged `release-1.2.0` on a `release/1.2.0` branch.
1. Run `./gradlew currentVersion` tasks and ensure it prints `1.2.0`
1. Make an additional commit
1. Run `./gradlew currentVersion` task again

**Expected result:**
Task completes with `1.2.1-SNAPSHOT` output, as it was in `1.12.1`

**Actual result:**
Task prints `1.3.0-SNAPSHOT` 

It looks like an oversight as a tag related function is used to set branch related property
https://github.com/allegro/axion-release-plugin/commit/de4ee1e8e78f3c97601efea8d46ae91a575d5644#diff-50b4d296a39fb72c511d8a25cd41aa5ccb5bc33637d078f6fbb61bf9158108d3L24

It seem like the workaround/fix is to set `releaseBranchPattern` explicitly, however I still haven't managed to make it work 👀 

In case when needed: [test failing on 1.13.1](https://github.com/usefulness/project-starter/blob/master/plugins/src/test/kotlin/com/project/starter/versioning/VersioningPluginTest.kt#L126) + plugin [config](https://github.com/usefulness/project-starter/blob/master/plugins/src/main/kotlin/com/project/starter/versioning/plugins/VersioningPlugin.kt#L25)

**Comment by bgalek** [comment]
I'll look at this tomorrow! Thanks for detailed issue submission!

**Comment by bgalek** [comment]
I've made a blind guess https://github.com/allegro/axion-release-plugin/pull/393/files (didn't test it yet) what do You think @mateuszkwiecinski ?

**Comment by bgalek** [comment]
@mateuszkwiecinski a quick "manual" test seems ok
```
➜  axion-release-plugin-test git:(main) git checkout release-0.1.2 
➜  axion-release-plugin-test git:(52c6208) gradle cV                 

> Task :currentVersion

Project version: 0.1.2

BUILD SUCCESSFUL in 845ms
1 actionable task: 1 executed
➜  axion-release-plugin-test git:(52c6208) git commit -am "test" --allow-empty 
[detached HEAD ce828f3] test
➜  axion-release-plugin-test git:(ce828f3) gradle cV                          

> Task :currentVersion

Project version: 0.1.3-SNAPSHOT

BUILD SUCCESSFUL in 599ms
1 actionable task: 1 executed
```

**Comment by bgalek** [comment]
@mateuszkwiecinski could you test it from this branch? I'll need some time to create a useful tests for this feature but also - I would like to fix broken release ;)

**Comment by mateuszkwiecinski** [comment]
Sure, I should be able to test it later today (somewhen around late evening), if not I'll do it as the very first thing tomorrow morning 👀 

Is there a snapshot version available or should I build and publish the plugin locally? 

Also, could you explain why it is expected to use tag config at release **branch** naming? Previously release branch format and tag name format were considered a separate things. AFAIU these are now unified. Is it expected to have branches named `v/1.2.3`? 

(or at least that's what I read from: `config.releaseBranchPattern = context.isLegacyDefTagnameRepo() ? TagPrefixConf.DEFAULT_LEGACY_PREFIX + '/.+' : TagPrefixConf.prefix() + '/.+'`)

According to what I read from the code it will work in my case and it'll use `release/1.2.3` branch, but it'll be broken for new users, and probably will work in an unexpected way if consumers will update tag prefix, no?

**Comment by bgalek** [comment]
There is no way (AFAIK) to have gradle plugins snapshots. So please use mavenLocal().

This plugin should not care about branch name - only tags matter.

Before my fix `incrementMinorIfNotOnRelease` incrementer assumed using default tag to check if it's a tagged commit or not (this was missed during 1.13.x version that changed default pattern to v1.2.3). 
So when using latest version, incrementer was looking for "v1.2.3" tag - but nowhere found (it was release-1.2.3), I've passed a flag to indicate is it old (release-) or new (v) and then decide. Of course, when somebody uses a custom prefix (ACME-) this should still work (incrementer would receive overridden version).

**Comment by bgalek** [comment]
https://axion-release-plugin.readthedocs.io/en/latest/configuration/version/#incrementminorifnotonrelease
Yes, the new default assumes v/1.2.3 branch naming pattern.

Before my fix **incrementMinorIfNotOnRelease** `releaseBranchPattern` argument used only new default value (v/1.2.3). I've passed a flag to indicate is it old default (release-) or new (v) and then decide. Of course, when somebody uses a custom `releaseBranchPattern` this should still work (incrementer would receive overridden version).

**Comment by bgalek** [comment]
Thanks for the idea with sonatype - it seems kinda obvious now as You mentioned it!
I use local maven for testing - don't see why I couldn't use external one 😅 

Please do check if everything seems all right on your end!

**Comment by mateuszkwiecinski** [comment]
@bgalek So: I managed to run tests linked in the original post, they all passed which means the legacy format is properly recognized now 👍 


As mentioned above, I had tried to set custom `releaseBranchPattern` but I couldn't make it work. 
Here's sample repro project: https://github.com/mateuszkwiecinski/axion_playground 

it has following repository state:
![image](https://user-images.githubusercontent.com/36954793/113197732-4dd3a400-9265-11eb-8b27-c8301ed62144.png)

and as you can see in the repository, my configuration looks as follows
_(Groovy)_
```groovy
scmVersion {
    tag {
        prefix = 'ohoh'
    }
    
    versionIncrementer = PredefinedVersionIncrementer.versionIncrementerFor('incrementMinorIfNotOnRelease')
    releaseBranchPattern = Pattern.compile('^random(/.*)?$')
}
```

I expected that the `IfNotOnRelease` part somehow relates to `releaseBranchPattern`, but clearly it doesn't and I should have used something like:
_(Kotlin)_
```kotlin
versionIncrementer = PredefinedVersionIncrementer.versionIncrementerFor("incrementMinorIfNotOnRelease", mapOf("releaseBranchPattern" to "^random(/.*)?$".toRegex().toPattern())
```

Is that expected? What's the purpose of `VersionConfig#releaseBranchPattern`? I couldn't find any usages of it 🤔

Other than that it seems to be working fine in all of my usecases (except #392)

**Comment by bgalek** [comment]
Than You for testing ang analysis! 

I'm gonna dig into `VersionConfig#releaseBranchPattern` docs clearly state: 
```
If incrementMinorIfNotOnRelease rule is activated then the property ``releaseBranchPattern`` is used to match the release
branch. By default it's set to 'release/.+'.
```
but it's from 2015 and maybe something broke. I'm wondering if before 1.13.x this was working?

**Comment by bgalek** [comment]
I've merged the fix as it preserves old functionality. I'll get back to this issue when confirmed that the bug was introduced in 1.13.x.

**Comment by mateuszkwiecinski** [comment]
>I'm wondering if before 1.13.x this was working?

It did not 😛 I tested on `1.11.0` and `1.12.1` and the `VersionConfig#releaseBranchPattern` option doesn't change the release branch pattern anywhere. 

I only started using it in `1.13.1` when I tried to work around the change introduced `1.13.x` with default naming convention (or actually to overcome this precise issue). 

I created separate issue: https://github.com/allegro/axion-release-plugin/issues/396 and I'm gona close this one as you fixed that already. Thank you 🙏
